lib/cog/command/pipeline/executor.ex

```diff
@@ -149,36 +117,51 @@ defmodule Cog.Command.Pipeline.Executor do
     end
   end
 
-  @doc """
-  `parse` -> {:stop, :shutdown} | {:next_state, :bind}
-
-  After connecting the request string is parsed and an invocation
-  (`%Piper.Ast.Invocation{}`) or a pipeline(`%Piper.Ast.Pipeline{}`) is added
-  to `remaining` in the state.
-  """
   def parse(:timeout, state) do
-    case Parser.scan_and_parse(state.request["text"], command_resolver: &CommandResolver.find_bundle(&1, state.user), return_pipeline: true) do
+    options = %ParserOptions{resolver: CommandResolver.command_resolver_fn(state.user)}
+    case Parser.scan_and_parse(state.request["text"], options) do
       {:ok, %Ast.Pipeline{}=pipeline} ->
-        {:next_state, :lookup_redirects, %{state | pipeline: pipeline}, 0}
+        # Extract and unwrap redirect information here; keeps AST
+        # details from creeping around, and eliminates the need to
+        # keep the pipeline around
+        redirect_destinations = case pipeline.redirect_to do
+                                  nil ->
+                                    []
+                                  %Ast.Redirect{targets: targets} ->
+                                    Enum.map(targets, &(&1.value))
+                                end
```

Comment:
Helper function on `Ast.Pipeline` would prevent the details of `Ast.Redirect` from leaking in here.

```
redirect_destinations = Ast.Pipeline.redirect_targets(pipeline) # Returns empty list or list of strings
```